src/main/java/com/flippingutilities/model/HistoryManager.java

```diff
@@ -621,9 +625,227 @@ else if (millisBetweenBuyAndSell < 0)
+		List<CombinationFlip> combinationFlips = getCombinationFlips();
+		//get only the combination flips that have all their offers after earliestTime
+		return combinationFlips.stream().
+				filter(cf -> cf.getOffers().stream().allMatch(po -> po.getOffer().getTime().isAfter(earliestTime))).
+				sorted(Comparator.comparing(cf -> cf.parent.getOffer().getTime())).
+				collect(Collectors.toList());
+	}
+
+	public List<CombinationFlip> getPersonalCombinationFlips(Instant earliestTime) {
+		return personalCombinationFlips.stream()
+				.filter(cf -> cf.getOffers().stream().allMatch(po -> po.getOffer().getTime().isAfter(earliestTime)))
+				.collect(Collectors.toList());
+	}
+
+	/**
+	 * Returns a view of the offer events that accounts for the amount consumed by partial offers
+	 * in the combination flips
+	 */
+	List<OfferEvent> getPartialOfferAdjustedView(List<OfferEvent> offers) {
+		Map<String, PartialOffer> offerIdsInCombinationFlips = getOfferIdToPartialOfferInComboFlips();
+		return offers.stream().map(o -> {
+			if (offerIdsInCombinationFlips.containsKey(o.getUuid())) {
+				return offerIdsInCombinationFlips.get(o.getUuid()).toAdjustedOfferEvent();
+			}
+			return o;
+		}).collect(Collectors.toList());
+	}
+
+	/**
+	 * @return all combination flips regardless of time
+	 */
+	private List<CombinationFlip> getCombinationFlips() {
+		List<CombinationFlip> combinationFlips = new ArrayList<>();
+		combinationFlips.addAll(personalCombinationFlips);
+		combinationFlips.addAll(parentCombinationFlips);
+		return combinationFlips;
+	}
+
+	/**
+	 * Because we don't persist the itemName field in an OfferEvent, we need to hydrate every
+	 * OfferEvent with it, which includes the offer events in combination flips
+	 * @param idToItemName a map of item id to item name for quick lookup
+	 */
+	public void setCombinationFlipOfferNames(Map<Integer, String> idToItemName) {
+		getCombinationFlips().forEach(cf -> cf.getOffers().forEach(po -> {
+			OfferEvent o = po.getOffer();
+			String itemName = idToItemName.getOrDefault(o.getItemId(), "");
+			o.setItemName(itemName);
+		}));
+	}
+
+	/**
+	 * Because we don't persist the itemName field in an OfferEvent, we need to hydrate every
+	 * OfferEvent with it
+	 */
+	public void setOfferNames(String itemName) {
+		compressedOfferEvents.forEach(o -> o.setItemName(itemName));
+	}
+
+	/**
+	 * We don't persist the madeBy field, so it has to be hydrated.
+	 */
+	public void setOfferMadeBy(String name) {
+		compressedOfferEvents.forEach(o -> o.setMadeBy(name));
+		getCombinationFlips().forEach(cf -> cf.getOffers().forEach(o -> o.getOffer().setMadeBy(name)));
+	}
+
+	/**
+	 * Every new offer event created nowadays has a uuid associated with it. However, the old
+	 * offer events that have already been persisted need their uuids' set.
+	 */
+	public void setOfferIds() {
+		compressedOfferEvents.forEach(o -> {
+			if (o.getUuid() == null) {
+				o.setUuid(UUID.randomUUID().toString());
+			}
+		});
```

Comment:
Most of the logic for the combination flips deals directly with the two fields in a HistoryManager that store CombinationFlips. If I put it in another component, I would have to ask the history manager for its combination flips and then perform operations on them. That wouldn't necessarily be bad, but since HistoryManager owns that data, I prefer if it performs the read/writes on it. Same with the offer events.